lib/vdc/BOVCollection.cpp

```diff
@@ -0,0 +1,368 @@
+    case DC::XType::DOUBLE: return 8;
+    default: return -1;
+    }
+}
+
+void BOVCollection::_swapBytes(void *vptr, size_t size, size_t n) const
+{
+    unsigned char *ucptr = (unsigned char *)vptr;
+    unsigned char  uc;
+    size_t         i, j;
+
+    for (j = 0; j < n; j++) {
+        for (i = 0; i < size / 2; i++) {
+            uc = ucptr[i];
+            ucptr[i] = ucptr[size - i - 1];
+            ucptr[size - i - 1] = uc;
+        }
+        ucptr += size;
+    }
+}
+
+template<class T> int BOVCollection::ReadRegion(const std::vector<size_t> &min, const std::vector<size_t> &max, T region)
+{
+    FILE *fp = fopen(_dataFile.c_str(), "rb");
+    if (!fp) {
+        SetErrMsg("Invalid file: %d", fp);
+        return -1;
+    }
+
+    size_t formatSize = _sizeOfFormat(_dataFormat);
+    if (formatSize < 0) {
+        SetErrMsg("Unspecified data format");
+        return (-1);
+    }
+
+    if (_dataSize.size() != 3) {
+        SetErrMsg("Invalid grid size (must be 3D)");
+        return (-1);
+    }
+    size_t numValues = _dataSize[0] * _dataSize[1] * _dataSize[2];
+
+    int  n = 1;
+    bool systemLittleEndian = *(char *)&n == 1 ? true : false;
+    bool dataLittleEndian = _dataEndian == "LITTLE" ? true : false;
+    bool needSwap = systemLittleEndian != dataLittleEndian ? true : false;
+
+    // Read a "pencil" of data along the X axis, one row at a time
+    size_t count = max[0] - min[0] + 1;
+    for (int k = min[2]; k <= max[2]; k++) {
+        int zOffset = _dataSize[0] * _dataSize[1] * k;
+        for (int j = min[1]; j <= max[1]; j++) {
+            int xOffset = min[0];
+            int yOffset = _dataSize[0] * j;
+            int offset = formatSize * (xOffset + yOffset + zOffset);
+
+            unsigned char readBuffer[count * formatSize];
+
+            if (needSwap) { _swapBytes(readBuffer, formatSize, numValues); }
+
+            fseek(fp, offset, SEEK_SET);
+            size_t rc = fread(readBuffer, formatSize, count, fp);
+
+            if (rc != count) {
+                if (ferror(fp) != 0) {
+                    MyBase::SetErrMsg("Error reading input file");
+                } else {
+                    MyBase::SetErrMsg("Short read on input file");
+                }
+                return -1;
+            }
+
+            if (_dataFormat == DC::XType::INT32) {
+                int *castBuffer = (int *)readBuffer;
+                for (int i = 0; i < count; i++) { *region++ = (typename std::remove_pointer<T>::type)castBuffer[i]; }
+            } else if (_dataFormat == DC::XType::FLOAT) {
+                float *castBuffer = (float *)readBuffer;
+                for (int i = 0; i < count; i++) { *region++ = (typename std::remove_pointer<T>::type)castBuffer[i]; }
+            } else if (_dataFormat == DC::XType::DOUBLE) {
+                double *castBuffer = (double *)readBuffer;
+                for (int i = 0; i < count; i++) { *region++ = (typename std::remove_pointer<T>::type)castBuffer[i]; }
```

Comment:
What is this code doing?

Reply:
This function takes a templated array and fills it.  I'm using std::remove_pointer to cast the 'unsigned char' values that are read from the file into the templated array.

I'm not sure if that's your specific question though.  If not, let me know.

Reply:
Is all of this necessary? Can't you simply declare vReadBuffer as type 'T'?

Reply:
For completeness, I'm adding your quoted text regarding this issue from another comment below.

> OK. My bad. You're having to do with two types: the format that the data are stored in the file in, and the format requested by the template.